src/dbnode/storage/shard.go

```diff
@@ -1765,16 +1766,24 @@ func (s *dbShard) FetchBlocksMetadataV2(
 				blockResult))
 		}
 
-		// Return the reader to the cache
 		endPos := int64(reader.MetadataRead())
-		s.namespaceReaderMgr.put(reader)
+		// This volume may be different from the one initialliy requested,
+		// e.g. if there was a compaction between the last call and this
+		// one.
```

Comment:
maybe add to the end of the comment: "so be sure to update the state of the pageToken otherwise the next request would have to start from the beginning since it would be requesting a stale volume and this could result in an infinite loop of requests that never complete." Or something like that.

Also a comment saying this must happen **before** the call to put would be nice just to make sure